File: 0047-permutations-ii/0047-permutations-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    nums.sort((a,b)=>{return a-b});
    const chosen = new Array(nums.length).fill(false);
    const permutation = [];
    const res = [];
    let prev = Number.MAX_INTEGER;
    const search = function () {
        if(permutation.length === nums.length) res.push(Array.from(permutation));
        else{
            for(let i = 0; i<nums.length; i++){
                if(chosen[i]) continue;
                const choice = nums[i];
                if(choice === prev) continue;
                else prev = Number.MAX_INTEGER;
                chosen[i] = true;
                permutation.push(choice);
                search();
                chosen[i] = false;
                prev = permutation.pop(); 
            }
        }
    }
    search();
    return res;
};

File: 0047-permutations-ii/0047-permutations-ii.cpp
class Solution {
private:
    unordered_map<int,int> mp;
    vector<int> permutation;
    vector<vector<int>> res;
public:
    void search(int n){
        if(permutation.size() == n){
            res.push_back(permutation);
        }else{
            for(auto& i : mp){
                if(i.second > 0){
                    permutation.push_back(i.first);
                    i.second--;
                    search(n);
                    i.second++;
                    permutation.pop_back();
                }
            }
        }
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        for(auto i : nums) mp[i]++;
        search(nums.size());
        return res;
    }
};